Neos.Fusion/Classes/Core/LegacyFusionControllerContext.php

```diff
@@ -0,0 +1,127 @@
+
+    public function __construct(
+        private readonly ActionRequest $request,
+        private readonly ActionResponse $legacyActionResponseForCurrentRendering
+    ) {
+    }
+
+    /**
+     * To migrate the use case of fetching the active request, please look into {@see FusionGlobals::get()} instead.
+     * By convention, an {@see ActionRequest} will be available as `request`:
+     *
+     *     $actionRequest = $this->runtime->fusionGlobals->get('request');
+     *     if (!$actionRequest instanceof ActionRequest) {
+     *         // fallback or error
+     *     }
+     *
+     * @deprecated with Neos 9.0 can be removed with 10
+     */
+    public function getRequest(): ActionRequest
+    {
+        return $this->request;
+    }
+
+    /**
+     * To migrate the use case of getting the UriBuilder please use this instead:
+     *
+     *     $actionRequest = $this->runtime->fusionGlobals->get('request');
+     *     if (!$actionRequest instanceof ActionRequest) {
+     *         // fallback or error
+     *     }
+     *     $uriBuilder = new UriBuilder();
+     *     $uriBuilder->setRequest($actionRequest);
+     *
+     * @deprecated with Neos 9.0 can be removed with 10
+     */
+    public function getUriBuilder(): UriBuilder
+    {
+        $uriBuilder = new UriBuilder();
+        $uriBuilder->setRequest($this->request);
+        return $uriBuilder;
+    }
+
+    /**
+     * To migrate this use case please use {@see FlashMessageService::getFlashMessageContainerForRequest()} in
+     * combination with fetching the active request as described here {@see getRequest} instead.
+     *
+     * @deprecated with Neos 9.0 can be removed with 10
+     */
+    public function getFlashMessageContainer(): FlashMessageContainer
+    {
+        return $this->flashMessageService->getFlashMessageContainerForRequest($this->request);
+    }
+
+    /**
+     * PURELY INTERNAL with partially undefined behaviour!!!
+     *
+     * Gives access to the legacy mutable action response simulation {@see Runtime::withSimulatedLegacyControllerContext()}
+     *
+     * Initially it was possible to mutate the current response of the active MVC controller though this getter.
+     *
+     * While *HIGHLY* internal behaviour and *ONLY* to be used by Neos.Fusion.Form or Neos.Neos:Plugin
+     * this legacy layer is in place still allows this functionality.
+     *
+     * @deprecated with Neos 9.0 can be removed with 10
+     * @internal THIS SHOULD NEVER BE CALLED ON USER-LAND
+     */
+    public function getResponse(): ActionResponse
+    {
+        // expose action response to be possibly mutated in neos forms or fusion plugins.
+        // this behaviour is highly internal and deprecated!
+        return $this->legacyActionResponseForCurrentRendering;
+    }
+
+    /**
+     * The method {@see ControllerContext::getArguments()} was removed without replacement.
+     */
+    // public function getArguments(): Arguments;
```

Comment:
it out-commented so in case someone looks up this code it will be clear that it was once part of the controller context but is not implemented here ;) Just a nice dev hint. I expect this class to be removed with Neos 10 or later so it doest have to be super pretty ^^